File: src/fwt/java/ScrollPanePeer.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   17 Jul 08  Brian Frank  Creation
//
package fan.fwt;

import fan.sys.*;
import fan.sys.List;
import org.eclipse.swt.*;
import org.eclipse.swt.widgets.*;
import org.eclipse.swt.widgets.Widget;
import org.eclipse.swt.widgets.ScrollBar;
import org.eclipse.swt.custom.ScrolledComposite;

public class ScrollPanePeer extends PanePeer
{

  public static ScrollPanePeer make(fan.fwt.ScrollPane self)
    throws Exception
  {
    ScrollPanePeer peer = new ScrollPanePeer();
    ((fan.fwt.Pane)self).peer = peer;
    ((fan.fwt.Widget)self).peer = peer;
    peer.self = self;
    return peer;
  }

  public Widget create(Widget parent)
  {
    ScrollPane self = (ScrollPane)this.self;

    int style = SWT.H_SCROLL | SWT.V_SCROLL;
    if (self.border)  style |= SWT.BORDER;

    ScrolledComposite c = new ScrolledComposite((Composite)parent, style);
    c.setExpandHorizontal(true);
    c.setExpandVertical(true);
    c.setMinSize(100,100);

    ScrollBar hbar = c.getHorizontalBar();
    ScrollBar vbar = c.getVerticalBar();
    if (hbar != null) ((ScrollBarPeer)self.hbar().peer).attachToScrollable(c, hbar);
    if (vbar != null) ((ScrollBarPeer)self.vbar().peer).attachToScrollable(c, vbar);

    this.control = c;
    return c;
  }

  void onSizeChange() { ((ScrollPane)self).onLayout(); }

  public void setMinSize(fan.fwt.ScrollPane self, fan.gfx.Size size)
  {
    ScrolledComposite sc = (ScrolledComposite)this.control;
    if (sc == null) return;
    sc.setMinSize((int)size.w, (int)size.h);
  }

}

File: src/fwt/js/ScrollPanePeer.js
//
// Copyright (c) 2010, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   04 Mar 10  Andy Frank  Creation
//

/**
 * ScrollPanePeer.
 */
fan.fwt.ScrollPanePeer = fan.sys.Obj.$extend(fan.fwt.PanePeer);
fan.fwt.ScrollPanePeer.prototype.$ctor = function(self)
{
  fan.fwt.PanePeer.prototype.$ctor.call(this, self);
}

fan.fwt.ScrollPanePeer.prototype.create = function(parentElem, self)
{
  var div = this.emptyDiv();
  div.style.overflow = "auto";
  parentElem.appendChild(div);
  return div;
}

fan.fwt.ScrollPanePeer.prototype.setMinSize = function(self, s)
{
  this.m_minSize = s;
}

fan.fwt.ScrollPanePeer.prototype.relayout = function(self)
{
  // short-circuit if not mounted
  if (this.elem == null) return;

  this.sync(self);
  if (self.onLayout) self.onLayout();

  var c = self.m_content;
  if (c != null)
  {
    var sz = this.m_size;
    var ms = this.m_minSize;
    var w = (sz.m_w < ms.m_w) ? ms.m_w : sz.m_w;
    var h = (sz.m_h < ms.m_h) ? ms.m_h : sz.m_h;
    c.size$(fan.gfx.Size.make(w,h));
    c.peer.relayout(c);
  }

  return self;
}